base/constructingfunctions.jl

```diff
@@ -0,0 +1,36 @@
+# Constructing functions for some concrete types.
+#
+# Meant to be used to improve inference, when the input arguments are not necessarily
+# precisely inferred. Partial workaround for issue #42372.
+module _ConstructingFunctions
+    export
+        _Bool,
+        _Int, _Int8, _Int16, _Int32, _Int64, _Int128,
+        _UInt, _UInt8, _UInt16, _UInt32, _UInt64, _UInt128,
+        _Char, _String
+    struct ConstructorStrict{F} <: Function end
+    # Keyword arguments are not available at this point in bootstrap, so they're not
+    # supported.
+    function (::ConstructorStrict{F})(args::Vararg{Any, N}) where {F, N}
+        @inline
+        r = F(args...)
+        r::F
+    end
```

Comment:
As a historical note, soon after closing this PR I realized there's a nicer way of doing this, based on preexisting functionality and more generic: `Base.Fix2(typeassert, F) ∘ F`. Registered package based on this idea:

* https://juliahub.com/ui/Packages/General/EnforcedTypeSignatureCallables

* https://github.com/JuliaFunctional/EnforcedTypeSignatureCallables.jl